### ISSUE TITLE
calibrating gamma for calibrated stacking

### ISSUE THREAD

**Issue by shadowbourne** [BatsResearch/nayak-arxiv20-code#10]
I was wondering if you could offer any advice on the following:

I am experimenting with different backbones and have noticed that γ=3 is no longer appropriate when you change away from resnet (on awa)

in the paper you say:
 > We calibrate γ on the validation splits as
suggested in Xian et al. (2018b) and set γ = 3.0 for AWA2
and γ = 2.0 for aPY for all the methods.

However I could find no reference of calibrated stacking being performed in Xian et al. (2018b)? 

Regardless, could you offer some advice on how exactly you calibrated gamma? Or alternatively point me towards where the code is within this repository, as I could not find it myself?

Thank you for your help!

**Comment by shadowbourne** [comment]
Was just following up on this as any help/clarification would be greatly appreciated as my thesis due date is approaching and this is a problem I am currently struggling on.

Thanks

**Comment by nihalnayak** [comment]
Apologies for the delay. The citation refers to the validation splits rather than the calibration. Equation 8 in the current version of the paper refers to the stacked calibration. As noted in the paper, our calibration is very similar to Xu et. al., 2020. 
You may get some insights after reading their work as well. 

The calibration coefficient is calibrated according to the validation splits mentioned in Xian et. al., 2018b. The validation splits for awa2 can be found at `object_classification/datasets/awa2` with the names `valclassesX.txt`. In our experiments, I fine-tuned the backbone on `trainclasses1.txt` and evaluated it on `valclasses1.txt`. You can fine-tune the backbone on different validation splits by changing `--fold` argument in `finetune_awa2.py`. If you choose 1, 2, or 3, the backbone fine-tunes on the different validation splits.

https://github.com/BatsResearch/nayak-arxiv20-code/blob/d7eced950ba1406d06c47bb664fcc01d6db7cf5c/object_classification/finetune_awa2.py#L42

During the evaluation, I varied the gamma coefficient and computed the harmonic mean of the macro averaged accuracy of the train and validation classes. I varied gamma from 0.0 to 5.0 with a 0.1 step size and evaluated the train and validation classes in `evaluate_awa2.py`. I choose the gamma value for the test set based on the best harmonic mean performance across the three test splits. I ensured that the test classes were still masked during the calibration. 

You will need to modify `evaluate_awa2.py` to calibrate the coefficient only on the train and validation classes and mask the test classes. I may not have the code that I used to compute the coefficient but this is pretty straightforward to code. 

I hope this helps! 

Best of luck with your thesis!
